File: frontend/src/services/workflowService.ts
import { apiClient } from './apiClient';
import type { ApiResponse } from './apiClient';

export interface TaskDTO {
  taskKey: number;
  taskType: string;
  taskId: string;
  processInstanceKey: number;
  processDefinitionId: string;
  assignee?: string;
  candidateUser?: string;
  candidateGroup?: string;
  creationTime: string;
  dueDate?: string;
  variables?: Record<string, any>;
  state: string;
}

// Alias for backward compatibility
export type WorkflowTask = TaskDTO;

export interface TaskCompleteRequest {
  taskId: number;
  variables?: Record<string, any>;
  comments?: string;
  approved?: boolean;
}

export interface ProcessInstanceDTO {
  processInstanceKey: number;
  processDefinitionKey: string;
  processDefinitionVersion: number;
  bpmnProcessId: string;
  state: string;
  startTime: string;
  endTime?: string;
  variables?: Record<string, any>;
}

export interface TaskHistoryDTO {
  taskKey: number;
  taskId: string;
  taskName: string;
  taskType: string;
  processInstanceKey: number;
  processDefinitionId: string;
  creationTime: string;
  createdBy?: string;
  assignmentTime?: string;
  assignedTo?: string;
  claimedTime?: string;
  claimedBy?: string;
  completionTime?: string;
  completedBy?: string;
  dueDate?: string;
  state: string;
  assignee?: string;
  candidateUser?: string;
  candidateGroup?: string;
  completionComment?: string;
  outcome?: string;
}

/**
 * Workflow service for managing workflow tasks and processes
 * Uses existing Workflow Adapter Service backend (port 8093)
 */
export const workflowService = {
  /**
   * Get tasks for a user
   */
  async getTasksForUser(userId: string): Promise<ApiResponse<TaskDTO[]>> {
    const response = await apiClient.get<ApiResponse<TaskDTO[]>>(
      `/api/v1/workflows/tasks/user/${userId}`
    );
    return response;
  },

  /**
   * Get tasks for a user (alias for backward compatibility)
   */
  async getUserTasks(userId: string): Promise<ApiResponse<TaskDTO[]>> {
    return this.getTasksForUser(userId);
  },

  /**
   * Get tasks by process instance
   */
  async getTasksByProcessInstance(processInstanceKey: number): Promise<ApiResponse<TaskDTO[]>> {
    const response = await apiClient.get<ApiResponse<TaskDTO[]>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}/tasks`
    );
    return response;
  },

  /**
   * Get process instance details
   */
  async getProcessInstance(processInstanceKey: number): Promise<ApiResponse<ProcessInstanceDTO>> {
    const response = await apiClient.get<ApiResponse<ProcessInstanceDTO>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}`
    );
    return response;
  },

  /**
   * Complete a task
   */
  async completeTask(taskId: number, variables?: Record<string, any>): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/complete`,
      variables || {}
    );
    return response;
  },

  /**
   * Get process instance history for a contract
   */
  async getProcessInstanceHistory(contractId: number): Promise<ApiResponse<any>> {
    const response = await apiClient.get<ApiResponse<any>>(
      `/api/v1/workflows/process-instances/contract/${contractId}/history`
    );
    return response;
  },

  /**
   * Start a process instance
   */
  async startProcess(processDefinitionKey: string, variables?: Record<string, any>): Promise<ApiResponse<ProcessInstanceResult>> {
    const response = await apiClient.post<ApiResponse<ProcessInstanceResult>>(
      `/api/v1/workflows/processes/${processDefinitionKey}/start`,
      variables || {}
    );
    return response;
  },

  /**
   * Cancel a process instance
   */
  async cancelProcessInstance(processInstanceKey: number): Promise<ApiResponse<void>> {
    const response = await apiClient.delete<ApiResponse<void>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}`
    );
    return response;
  },

  /**
   * Assign a task to a user
   */
  async assignTask(taskId: number, userId: string): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/assign`,
      { userId }
    );
    return response;
  },

  /**
   * Claim an unassigned task
   */
  async claimTask(taskId: number, userId: string): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/claim`,
      { userId }
    );
    return response;
  },

  /**
   * Unclaim a task
   */
  async unclaimTask(taskId: number): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/unclaim`
    );
    return response;
  },

  /**
   * Reassign a task to a different user
   */
  async reassignTask(taskId: number, userId: string): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/reassign`,
      { userId }
    );
    return response;
  },

  /**
   * Return a task to the previous step
   */
  async returnTask(taskId: number, reason?: string): Promise<ApiResponse<void>> {
    const response = await apiClient.post<ApiResponse<void>>(
      `/api/v1/workflows/tasks/${taskId}/return`,
      reason ? { reason } : {}
    );
    return response;
  },

  /**
   * Get BPMN XML for a process definition
   * Framework-based: works for any workflow
   */
  async getBpmnXml(processDefinitionKey: string): Promise<ApiResponse<string>> {
    const response = await apiClient.get<ApiResponse<string>>(
      `/api/v1/workflows/process-definitions/${processDefinitionKey}/bpmn`
    );
    return response;
  },

  /**
   * Get flow node instances (activities) for a process instance
   * Used for visualizing workflow status in BPMN diagrams
   */
  async getFlowNodeInstances(processInstanceKey: number): Promise<ApiResponse<ActivityInstanceDTO[]>> {
    const response = await apiClient.get<ApiResponse<ActivityInstanceDTO[]>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}/activities`
    );
    return response;
  },

  /**
   * Get incidents (errors) for a process instance
   * Used for visualizing errors in BPMN diagrams
   */
  async getIncidents(processInstanceKey: number): Promise<ApiResponse<IncidentDTO[]>> {
    const response = await apiClient.get<ApiResponse<IncidentDTO[]>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}/incidents`
    );
    return response;
  },

  /**
   * Get task history for a process instance
   * Returns detailed timeline of task creation, assignment, and completion/approval
   */
  async getTaskHistory(processInstanceKey: number): Promise<ApiResponse<TaskHistoryDTO[]>> {
    const response = await apiClient.get<ApiResponse<TaskHistoryDTO[]>>(
      `/api/v1/workflows/process-instances/${processInstanceKey}/task-history`
    );
    return response;
  },
};

export interface ProcessInstanceResult {
  processInstanceKey: number;
  processDefinitionKey: string;
  processDefinitionVersion: number;
  bpmnProcessId: string;
  variables?: Record<string, any>;
}

export interface ActivityInstanceDTO {
  activityInstanceKey: number;
  activityId: string;
  activityName: string;
  activityType: string;
  state: string; // COMPLETED, ACTIVE, TERMINATED
  startTime?: string;
  endTime?: string;
  assignee?: string;
}

export interface IncidentDTO {
  incidentKey: number;
  incidentType: string;
  errorMessage: string;
  creationTime?: string;
  resolutionTime?: string;
  state: string; // OPEN, RESOLVED
}

File: backend/workflow-adapter-service/src/main/java/com/hhsa/workflow/service/WorkflowService.java
package com.hhsa.workflow.service;

import com.hhsa.common.core.dto.ApiResponse;
import com.hhsa.workflow.adapter.WorkflowAdapter;
import com.hhsa.workflow.adapter.WorkflowException;
import com.hhsa.workflow.dto.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

/**
 * Workflow service providing high-level workflow operations.
 * Delegates to WorkflowAdapter for actual implementation.
 */
@Service
public class WorkflowService {

    private static final Logger logger = LoggerFactory.getLogger(WorkflowService.class);

    private final WorkflowAdapter workflowAdapter;
    private final BpmnDeploymentService bpmnDeploymentService;

    public WorkflowService(WorkflowAdapter workflowAdapter, BpmnDeploymentService bpmnDeploymentService) {
        this.workflowAdapter = workflowAdapter;
        this.bpmnDeploymentService = bpmnDeploymentService;
    }

    /**
     * Deploy BPMN process
     */
    public DeploymentResult deployProcess(String bpmnXml) throws WorkflowException {
        logger.debug("Deploying BPMN process");
        return workflowAdapter.deployProcess(bpmnXml);
    }

    /**
     * Start process instance
     */
    public ProcessInstanceResult startProcess(String processDefinitionKey, Map<String, Object> variables) throws WorkflowException {
        logger.debug("Starting process: {}", processDefinitionKey);
        return workflowAdapter.startProcess(processDefinitionKey, variables);
    }

    /**
     * Complete task
     */
    public void completeTask(Long taskId, Map<String, Object> variables) throws WorkflowException {
        logger.debug("Completing task: {}", taskId);
        workflowAdapter.completeTask(taskId, variables);
    }

    /**
     * Get tasks for user
     */
    public List<TaskDTO> getTasksForUser(String userId) throws WorkflowException {
        logger.debug("Getting tasks for user: {}", userId);
        return workflowAdapter.getTasksForUser(userId);
    }

    /**
     * Get tasks by process instance
     */
    public List<TaskDTO> getTasksByProcessInstance(Long processInstanceKey) throws WorkflowException {
        logger.debug("Getting tasks for process instance: {}", processInstanceKey);
        return workflowAdapter.getTasksByProcessInstance(processInstanceKey);
    }

    /**
     * Get process instance
     */
    public ProcessInstanceDTO getProcessInstance(Long processInstanceKey) throws WorkflowException {
        logger.debug("Getting process instance: {}", processInstanceKey);
        return workflowAdapter.getProcessInstance(processInstanceKey);
    }

    /**
     * Cancel process instance
     */
    public void cancelProcessInstance(Long processInstanceKey) throws WorkflowException {
        logger.debug("Cancelling process instance: {}", processInstanceKey);
        workflowAdapter.cancelProcessInstance(processInstanceKey);
    }

    /**
     * Evaluate DMN decision
     */
    public DecisionResult evaluateDecision(String decisionId, Map<String, Object> variables) throws WorkflowException {
        logger.debug("Evaluating DMN decision: {}", decisionId);
        return workflowAdapter.evaluateDecision(decisionId, variables);
    }

    /**
     * Get process instance history for a contract
     */
    public ProcessInstanceHistoryDTO getProcessInstanceHistory(Long contractId) throws WorkflowException {
        logger.debug("Getting process instance history for contract: {}", contractId);
        return workflowAdapter.getProcessInstanceHistory(contractId);
    }

    /**
     * Get flow node instances (activities) for a process instance
     * Used for visualizing workflow status in BPMN diagrams
     */
    public List<ActivityInstanceDTO> getFlowNodeInstances(Long processInstanceKey) throws WorkflowException {
        try {
            logger.debug("Getting flow node instances for process instance: {}", processInstanceKey);
            List<ActivityInstanceDTO> result = workflowAdapter.getFlowNodeInstances(processInstanceKey);
            return result != null ? result : new java.util.ArrayList<>();
        } catch (Exception e) {
            logger.warn("Failed to get flow node instances for process instance: {} - {}", processInstanceKey, e.getMessage());
            // Return empty list for graceful degradation
            return new java.util.ArrayList<>();
        }
    }

    /**
     * Get incidents (errors) for a process instance
     * Used for visualizing errors in BPMN diagrams
     */
    public List<IncidentDTO> getIncidents(Long processInstanceKey) throws WorkflowException {
        try {
            logger.debug("Getting incidents for process instance: {}", processInstanceKey);
            List<IncidentDTO> result = workflowAdapter.getIncidents(processInstanceKey);
            return result != null ? result : new java.util.ArrayList<>();
        } catch (Exception e) {
            logger.warn("Failed to get incidents for process instance: {} - {}", processInstanceKey, e.getMessage());
            // Return empty list for graceful degradation
            return new java.util.ArrayList<>();
        }
    }

    /**
     * Get task history for a process instance
     * Returns detailed timeline of task creation, assignment, and completion/approval
     */
    public List<com.hhsa.workflow.dto.TaskHistoryDTO> getTaskHistory(Long processInstanceKey) throws WorkflowException {
        try {
            logger.debug("Getting task history for process instance: {}", processInstanceKey);
            if (workflowAdapter instanceof com.hhsa.workflow.adapter.Camunda8WorkflowAdapter) {
                com.hhsa.workflow.adapter.Camunda8WorkflowAdapter camundaAdapter = 
                    (com.hhsa.workflow.adapter.Camunda8WorkflowAdapter) workflowAdapter;
                return camundaAdapter.getTaskHistory(processInstanceKey);
            }
            return new java.util.ArrayList<>();
        } catch (Exception e) {
            logger.warn("Failed to get task history for process instance: {} - {}", processInstanceKey, e.getMessage());
            return new java.util.ArrayList<>();
        }
    }

    /**
     * Get BPMN XML for a process definition
     * Framework-based: works for any workflow by process definition key
     */
    public String getBpmnXml(String processDefinitionKey) throws WorkflowException {
        logger.debug("Getting BPMN XML for process definition: {}", processDefinitionKey);
        return bpmnDeploymentService.getBpmnXml(processDefinitionKey);
    }

    /**
     * Assign a task to a user
     */
    public void assignTask(Long taskId, String userId) throws WorkflowException {
        logger.debug("Assigning task: {} to user: {}", taskId, userId);
        workflowAdapter.assignTask(taskId, userId);
    }

    /**
     * Claim an unassigned task
     */
    public void claimTask(Long taskId, String userId) throws WorkflowException {
        logger.debug("Claiming task: {} by user: {}", taskId, userId);
        workflowAdapter.claimTask(taskId, userId);
    }

    /**
     * Unclaim a task
     */
    public void unclaimTask(Long taskId) throws WorkflowException {
        logger.debug("Unclaiming task: {}", taskId);
        workflowAdapter.unclaimTask(taskId);
    }

    /**
     * Reassign a task to a different user
     */
    public void reassignTask(Long taskId, String userId) throws WorkflowException {
        logger.debug("Reassigning task: {} to user: {}", taskId, userId);
        workflowAdapter.reassignTask(taskId, userId);
    }

    /**
     * Return a task to the previous step
     */
    public void returnTask(Long taskId, String reason) throws WorkflowException {
        logger.debug("Returning task: {} with reason: {}", taskId, reason);
        workflowAdapter.returnTask(taskId, reason);
    }
}

